Distributed Transaction Routing (DTR)

cwilvx opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Implemented Distributed Transaction Routing (DTR) system for optimized transaction relay across validators
  * Added background retry service to automatically reprocess and relay pending transactions at intervals
  * Introduced validator-aware transaction routing with deterministic validator selection and load balancing
  * Enhanced transaction relay with caching and multi-attempt retry logic for improved reliability

* **Chores**
  * Updated repository ignore patterns for configuration and key files
  * Cleaned up unused imports across consensus modules

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
ignored sensitive files

Commit:
dtr first test

Commit:
Merge branch 'testnet' into dtr

Commit:
updated docs

Commit:
entrypoint for background tx relaying

Commit:
added deletion method

Commit:
hardened DTR logic

Commit:
relayRetryService implementation for background relay of txs

cwilvx committed:
merge branch 'testnet' into dtr

cwilvx committed:
move all dtr functions to DTRManager

+ fix relax_tx nodecall call payload parameters
+ add signature verification to relay_tx handler
+ add node identitification to relay_tx handler
+ remove transactio object from relax_tx payload
+ rename relay service to dtr manager
+ use promise.allSettled to dispatch tx relays
+ return validators from isValidatorForNextBlock helper
+ add signing algorithm checks for relayer node

coderabbitai[bot] committed:
📝 Add docstrings to `dtr`

Docstrings generation was requested by @tcsenpai.

* https://github.com/kynesyslabs/node/pull/517#issuecomment-3616750620

The following files were modified:

* `src/index.ts`
* `src/libs/consensus/v2/routines/getShard.ts`
* `src/libs/consensus/v2/routines/isValidator.ts`
* `src/libs/network/manageNodeCall.ts`

tcsenpai committed:
Merge pull request #521 from kynesyslabs/coderabbitai/docstrings/b8f4d11

📝 Add docstrings to `dtr`

Commit:
Sync AGENTS.md from testnet

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

Commit:
Merge remote-tracking branch 'refs/remotes/origin/dtr' into dtr

Commit:
beads cleanup

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

Commit:
updated beads memories

cwilvx committed:
handle tx relay while in consensus loop

+ fix: synced blocks having no transactions
+ update RELAY_TX nodecall to accept a list
+ add dtr waiter key
+ refactor relayRetryService to DTR Manager
+ put back sleep 250ms in requestBlocks while loop (sync.ts)
+ start waitForBlockThenRelay if tx comes in in consensus loop
+ release dtr waiter after finalize block
+

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
fix: transaction sync missing transactions

+ add new getBlockTransactions node call
+ return null if rawTx is nullish (fromRawTransaction)
+ use getBlockTransactions to download block transactions in sync.ts
+ try/catch in for loops when running sync gcr edits and tx inserts
+ delete relay cache entries after consuming them in rpc node
+ more logging

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
add getBlockTransactions node call handler

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
update merge mempool step to request pool difference between validators

+ release dtr relay waiter after consensus
+ bump shard size back to 4 nodes

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
revert merge mempool using difference

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
try: enforce blockref for relayed transactions

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
try: retry block sync after failure using while loop

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
use strict block number match

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->

cwilvx committed:
cleanup

tcsenpai commented:
## Your trial has ended! 😢

To keep getting reviews, activate your plan [here](https://app.kodus.io/settings/subscription).

Got questions about plans or want to see if we can extend your trial? Talk to our founders [here](https://cal.com/gabrielmalinosqui/30min).😎

<!-- kody-codereview -->
